SystemClockChangeTest failures [HZ-1834]

**testCluster_whenSlaveClockJumpsForward**

```
java.lang.AssertionError: Cluster should be stable when system clock changes! expected:<2> but was:<1>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:645)
    at com.hazelcast.cluster.SystemClockChangeTest.assertClusterSizeAlways(SystemClockChangeTest.java:132)
    at com.hazelcast.cluster.SystemClockChangeTest.testCluster_whenSlaveClockJumpsForward(SystemClockChangeTest.java:99)
```

**testCluster_whenMasterClockJumpsForward**

```
java.lang.AssertionError: Cluster time should be equal to master time! expected:<1.453851074083E12> but was:<1.453849274088E12>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:553)
    at com.hazelcast.cluster.SystemClockChangeTest.assertClusterTime(SystemClockChangeTest.java:125)
    at com.hazelcast.cluster.SystemClockChangeTest.testCluster_whenMasterClockJumpsForward(SystemClockChangeTest.java:77)
```

**testCluster_whenSlaveClockJumpsBackward**

```
java.lang.AssertionError: Cluster should be stable when system clock changes! expected:<2> but was:<1>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:645)
    at com.hazelcast.cluster.SystemClockChangeTest.assertClusterSizeAlways(SystemClockChangeTest.java:132)
    at com.hazelcast.cluster.SystemClockChangeTest.testCluster_whenSlaveClockJumpsBackward(SystemClockChangeTest.java:110)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/844/

**testCluster_whenSlaveClockJumpsForward**
```
java.lang.AssertionError: Cluster should be stable when system clock changes! expected:<2> but was:<1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.cluster.SystemClockChangeTest.assertClusterSizeAlways(SystemClockChangeTest.java:132)
	at com.hazelcast.cluster.SystemClockChangeTest.testCluster_whenSlaveClockJumpsForward(SystemClockChangeTest.java:99)
```
**testCluster_whenMasterClockJumpsBackward**
```
java.lang.AssertionError: Cluster time should be equal to master time! expected:<1.455568626108E12> but was:<1.455566826113E12>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:553)
	at com.hazelcast.cluster.SystemClockChangeTest.assertClusterTime(SystemClockChangeTest.java:125)
	at com.hazelcast.cluster.SystemClockChangeTest.testCluster_whenMasterClockJumpsBackward(SystemClockChangeTest.java:89)
```
**estCluster_whenSlaveClockJumpsBackward**
```
java.lang.AssertionError: Cluster should be stable when system clock changes! expected:<2> but was:<1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.cluster.SystemClockChangeTest.assertClusterSizeAlways(SystemClockChangeTest.java:132)
	at com.hazelcast.cluster.SystemClockChangeTest.testCluster_whenSlaveClockJumpsBackward(SystemClockChangeTest.java:110)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.maintenance-nightly/763/

**testCluster_whenSlaveClockJumpsBackward**
```
java.lang.AssertionError: Cluster time should be (approx.) equal to master time! expected:<1.456280752582E12> but was:<1.456282552586E12>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:553)
	at com.hazelcast.cluster.SystemClockChangeTest.assertClusterTime(SystemClockChangeTest.java:183)
	at com.hazelcast.cluster.SystemClockChangeTest.testCluster_whenSlaveClockJumpsBackward(SystemClockChangeTest.java:174)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/872/testReport/junit/com.hazelcast.cluster/SystemClockChangeTest/testCluster_whenSlaveClockJumpsBackward/

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterHeartbeatManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterHeartbeatManager.java
@@ -85,6 +85,7 @@ public class ClusterHeartbeatManager {
 
     @Probe(name = "lastHeartBeat")
     private volatile long lastHeartBeat;
+    private volatile long lastClusterTimeDiff;
 
     public ClusterHeartbeatManager(Node node, ClusterServiceImpl clusterService) {
         this.node = node;
@@ -199,8 +200,14 @@ public class ClusterHeartbeatManager {
                 logger.info(format("System clock apparently jumped from %s to %s since last heartbeat (%+d ms)",
                         sdf.format(new Date(lastHeartBeat)), sdf.format(new Date(now)), clockJump));
 
-                // adjust cluster clock due to clock drift
-                clusterClock.setClusterTimeDiff(clusterClock.getClusterTimeDiff() - clockJump);
+                // We only set cluster clock, if clock jumps more than threshold.
+                // If the last cluster-time diff we've seen is significantly different than what we read now,
+                // that means, it's already adjusted by master heartbeat. Then don't update the cluster time again.
+                long currentClusterTimeDiff = clusterClock.getClusterTimeDiff();
+                if (Math.abs(lastClusterTimeDiff - currentClusterTimeDiff) < CLOCK_JUMP_THRESHOLD) {
+                    // adjust cluster clock due to clock drift
+                    clusterClock.setClusterTimeDiff(currentClusterTimeDiff - clockJump);
+                }
             }
 
             if (absoluteClockJump >= maxNoMasterConfirmationMillis / 2) {
@@ -214,6 +221,7 @@ public class ClusterHeartbeatManager {
                 resetHeartbeats();
             }
         }
+        lastClusterTimeDiff = clusterClock.getClusterTimeDiff();
         lastHeartBeat = now;
     }
 

